app/plugin.go

```diff
@@ -259,13 +255,35 @@ func (a *App) SyncPlugins() *model.AppError {
 }
 
 func (a *App) ShutDownPlugins() {
-	pluginsEnvironment := a.GetPluginsEnvironment()
+	// Acquiring lock manually, as plugins might be disabled. See GetPluginsEnvironment.
+	a.Srv.PluginsLock.RLock()
+	pluginsEnvironment := a.Srv.PluginsEnvironment
+	a.Srv.PluginsLock.RUnlock()
+
 	if pluginsEnvironment == nil {
+		mlog.Warn("Plugins are already shutdown while trying to shut them down")
 		return
 	}
 
 	mlog.Info("Shutting down plugins")
 
+	availablePlugins, err := pluginsEnvironment.Available()
+	if err != nil {
+		mlog.Error("Failed to get available plugins while trying to shut them down. Clients will not get notified about shut down plugins.", mlog.Err(err))
+	}
+	for _, plugin := range availablePlugins {
```

Comment:
But doesn't this happen when any server in HA shuts down? You shut down one server and all your client plugins are disabled?

Reply:
Good catch! Fixed.

Tested this case with the new code and can confirm that this doesn't happen now.

Reply:
Good catch! Fixed.

Tested this case with the new code and can confirm that this doesn't happen now.